File: apps/web/dev-server.mjs
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { extname, join, normalize } from "node:path";

const root = new URL(".", import.meta.url).pathname;
const repoRoot = new URL("../../", import.meta.url).pathname;
const port = Number(process.env.PORT || 5173);

const mime = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8"
};

function resolvePath(urlPath) {
  if (urlPath.startsWith("/data/")) {
    return join(repoRoot, normalize(urlPath));
  }
  const path = urlPath === "/" ? "/index.html" : urlPath;
  return join(root, normalize(path));
}

createServer(async (req, res) => {
  try {
    const url = new URL(req.url || "/", `http://${req.headers.host}`);
    const file = resolvePath(url.pathname);
    const body = await readFile(file);
    res.writeHead(200, { "content-type": mime[extname(file)] || "application/octet-stream" });
    res.end(body);
  } catch (error) {
    res.writeHead(404, { "content-type": "text/plain; charset=utf-8" });
    res.end(`Not found: ${req.url}`);
  }
}).listen(port, "127.0.0.1", () => {
  console.log(`Web dashboard: http://127.0.0.1:${port}`);
});

File: apps/web/dev_server.py
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
import os

REPO_ROOT = Path(__file__).resolve().parents[2]
WEB_ROOT = Path(__file__).resolve().parent
PORT = int(os.environ.get("PORT", "5173"))


class Handler(SimpleHTTPRequestHandler):
    def translate_path(self, path):
        if path.startswith("/data/"):
            return str(REPO_ROOT / path.lstrip("/"))
        return str(WEB_ROOT / ("index.html" if path == "/" else path.lstrip("/")))


if __name__ == "__main__":
    server = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    print(f"Web dashboard: http://127.0.0.1:{PORT}")
    server.serve_forever()

